api/v2/Routes/students/masterymapping/index.js

```diff
@@ -0,0 +1,75 @@
+import { Router } from 'express';
+import {
+    getMaxScores,
+    getStudentScores,
+} from '../../../../lib/redisHelper.mjs';
+import ProgressReportData from '../../../../assets/progressReport/CS10.json' assert { type: 'json' };
+
+const router = Router({ mergeParams: true });
+
+function getTopicsFromUser(gradeData) {
+    const topicsTable = {};
+    Object.entries(gradeData)
+        .flatMap(([assignment, topics]) => Object.entries(topics))
+        .forEach(([topic, score]) => {
+            if (topic in topicsTable) {
+                topicsTable[topic] += +(score ?? 0);
+            } else {
+                topicsTable[topic] = +(score ?? 0);
+            }
+        });
+    return topicsTable;
+}
+
+async function getMasteryMapping(userTopicPoints, maxTopicPoints) {
+    const numMasteryLevels = ProgressReportData['student levels'].length - 2;
+    Object.entries(userTopicPoints).forEach(([topic, userPoints]) => {
+        const maxAchievablePoints = maxTopicPoints[topic];
+        if (userPoints === 0) {
+            return;
+        }
+        if (userPoints >= maxAchievablePoints) {
+            userTopicPoints[topic] = numMasteryLevels + 1;
+            return;
+        }
+        const unBoundedMasteryLevel =
+            (userPoints / maxAchievablePoints) * numMasteryLevels;
+        if (unBoundedMasteryLevel === numMasteryLevels) {
+            userTopicPoints[topic] = numMasteryLevels;
+        } else if (unBoundedMasteryLevel % 1 === 0) {
+            // Push them over to the next category if they are exactly on the edge.
+            userTopicPoints[topic] = unBoundedMasteryLevel + 1;
+        } else {
+            userTopicPoints[topic] = Math.ceil(unBoundedMasteryLevel);
+        }
+    });
+    const masteryMapping = {};
+    Object.entries(userTopicPoints).forEach(([topic, userPoints]) => {
+        masteryMapping[topic] = {"student_mastery": userPoints, "class_mastery": 0};
+    });
+    return masteryMapping;
+}
+
+router.get('/', async (req, res) => {
+    const { id } = req.params;
+    try {
+        const maxScores = await getMaxScores();
+        const studentScores = await getStudentScores(id);
+        const userTopicPoints = getTopicsFromUser(studentScores);
+        const maxTopicPoints = getTopicsFromUser(maxScores);
+        const masteryNum = await getMasteryMapping(userTopicPoints, maxTopicPoints);
```

Comment:
nit(perf): would be good to use `Promise.all()` to parallelize, but this is a total nit, so you can just ignore and we can refactor later.

Reply:
Noted!